### PR TITLE
Upgrade tonic, prost, otel to latest

### PR DESCRIPTION
**Pull request by h7kanna**
Upgrade Tonic, Prost, OTel deps

## What was changed
Updated dependencies
tonic - 0.12.x
prost - 0.13.x
otel - 0.24.x

## Why?
To upgrade to hyper 1.x ecosystem

## Checklist

1. Closes [#626](https://github.com/temporalio/sdk-core/issues/626)

3. How was this tested:
```
cargo test
cargo integ-test

```
To avoid mixing changes (Rust 1.80 clippy throwing lints)
```
cargo clippy --all -- -D warnings -A clippy::clone-on-copy
```

4. Any docs updates needed?
NA

### CODE REVIEW

**Review by Sushisource — COMMENTED** [review]
Thanks for this! Glad the stuff is finally updated enough that everything can use the newer versions.

Looks like there is a clippy lint that needs to be fixed, and the functionality of the connection override needs to be restored.

File: client/src/proxy.rs [hunk 1]
```diff
@@ -43,12 +46,13 @@ impl HttpConnectProxyOptions {
...
-        let req = req_build.body(hyper::Body::empty())?;
+        let req = req_build.body(Empty::<Bytes>::new())?;
 
         // We have to create a client with a specific connector because Hyper is
         // not letting us change the HTTP/2 authority
-        let client =
-            hyper::Client::builder().build(OverrideAddrConnector(self.target_addr.clone()));
+        // let client = hyper::Client::builder().build(OverrideAddrConnector(self.target_addr.clone()));
+
+        let client = Client::builder(TokioExecutor::new()).build_http();
```
[the remark this answers is not in the document]
Comment (h7kanna) [reply client/src/proxy.rs:55]: Oh missed it, working on it.

**Comment by h7kanna** [comment]
> Thanks for this! Glad the stuff is finally updated enough that everything can use the newer versions.
> 
> Looks like there is a clippy lint that needs to be fixed, and the functionality of the connection override needs to be restored.

There are lot of clippy lints about clone-with-copy, should I fix this in this PR?

**Comment by Sushisource** [comment]
> > Thanks for this! Glad the stuff is finally updated enough that everything can use the newer versions.
> > Looks like there is a clippy lint that needs to be fixed, and the functionality of the connection override needs to be restored.
> 
> There are lot of clippy lints about clone-with-copy unrelated to these updates, should I fix this in this PR? I can put up another separate PR along with this.

At least this one looks related, looks like the timestamp type became copyable and hence is triggering this now. If there are also other, unrelated ones, fixing them in this PR is totally fine.

File: client/src/proxy.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (h7kanna) [reply client/src/proxy.rs]: Looks like not as straightforward as before with hyper 1.x. 
Took a first stab it. But needs more understanding.

**Comment by h7kanna** [comment]
> > > Thanks for this! Glad the stuff is finally updated enough that everything can use the newer versions.
> > > Looks like there is a clippy lint that needs to be fixed, and the functionality of the connection override needs to be restored.
> > 
> > 
> > There are lot of clippy lints about clone-with-copy unrelated to these updates, should I fix this in this PR? I can put up another separate PR along with this.
> 
> At least this one looks related, looks like the timestamp type became copyable and hence is triggering this now. If there are also other, unrelated ones, fixing them in this PR is totally fine.

Done fixed

**Review by Sushisource — APPROVED** [review]

**Pull request merged**